File: qovis_backend/entity/transform.py
from enum import Enum
from typing import Optional


class TransType(Enum):
    INSERT = 'insert'
    DELETE = 'delete'
    REPLACE = 'replace'
    MODIFY = 'modify'
    UNCHANGE = 'unchange'


class Transform:
    def __init__(self, src_vid: Optional[int], dst_vid: Optional[int], type_: TransType):
        self.src_vid: Optional[int] = src_vid
        self.dst_vid: Optional[int] = dst_vid
        self.type: TransType = type_

    def dump(self):
        return {
            'srcVid': self.src_vid,
            'dstVid': self.dst_vid,
            'type': self.type.value,
        }

    @staticmethod
    def load(data):
        return Transform(
            src_vid=data['srcVid'],
            dst_vid=data['dstVid'],
            type_=TransType(data['type']),
        )

    def __repr__(self):
        return f'{self.type.value} {self.src_vid} -> {self.dst_vid}'

File: qovis_frontend/src/js/probe/Transform.js
import {VisTransform} from "@/js/probe/VisTransform"

/**
 * @enum {string}
 * @readonly
 */
export const TransType = {
  INSERT: 'insert',
  DELETE: 'delete',
  REPLACE: 'replace',
  MODIFY: 'modify',
  UNCHANGE: 'unchange',
}

export class Transform {
  /** @type {VisTransform} */
  visObj = null

  srcVid
  dstVid
  /** @type {TransType} */
  type

  /**
   * @param {int | null} srcVid
   * @param {int | null} dstVid
   * @param {TransType} type
   */
  constructor(srcVid, dstVid, type) {
    this.srcVid = srcVid
    this.dstVid = dstVid
    this.type = type
  }

  createVisObj() {
    this.visObj = new VisTransform(this)
    return this.visObj
  }

  static load(data) {
    return new Transform(
      data.srcVid,
      data.dstVid,
      data.type,
    )
  }

  toString() {
    return `${this.type} ${this.srcVid} -> ${this.dstVid}`
  }
}
